src/py/mat3ra/made/tools/build/passivation/builders.py

```diff
@@ -0,0 +1,183 @@
+        depth (float): Depth from the top to look for exposed surface atoms to passivate, in Angstroms.
+    """
+
+    shadowing_radius: float = 2.5
+    depth: float = 5.0
+
+
+class SurfacePassivationBuilder(PassivationBuilder):
+    """
+    Builder for passivating a surface.
+
+    Detects surface atoms looking along Z axis and passivates either the top or bottom surface or both.
+    """
+
+    build_parameters: SurfacePassivationBuilderParameters = SurfacePassivationBuilderParameters()
+    _ConfigurationType = PassivationConfiguration
+
+    def create_passivated_material(self, configuration: PassivationConfiguration) -> Material:
+        material = super().create_passivated_material(configuration)
+        passivant_coordinates_values = []
+
+        if configuration.surface in (SurfaceTypes.BOTTOM, SurfaceTypes.BOTH):
+            passivant_coordinates_values.extend(
+                self._get_passivant_coordinates(material, SurfaceTypes.BOTTOM, configuration)
+            )
+        if configuration.surface in (SurfaceTypes.TOP, SurfaceTypes.BOTH):
+            passivant_coordinates_values.extend(
+                self._get_passivant_coordinates(material, SurfaceTypes.TOP, configuration)
+            )
+
+        return self._add_passivant_atoms(material, passivant_coordinates_values, configuration.passivant)
+
+    def _get_passivant_coordinates(
+        self, material: Material, surface: SurfaceTypes, configuration: PassivationConfiguration
+    ):
+        """
+        Calculate the coordinates for placing passivants based on the specified surface type.
+
+        Args:
+            material (Material): Material to passivate.
+            surface (SurfaceTypes): Surface type (TOP or BOTTOM).
+            configuration (SurfacePassivationConfiguration): Configuration for passivation.
+
+        Returns:
+            list: Coordinates where passivants should be added.
+        """
+        surface_atoms_indices = get_surface_atom_indices(
+            material, surface, self.build_parameters.shadowing_radius, self.build_parameters.depth
+        )
+        surface_atoms_coordinates = [
+            material.basis.coordinates.get_element_value_by_index(i) for i in surface_atoms_indices
+        ]
+        bond_vector = (
+            [0, 0, configuration.bond_length] if surface == SurfaceTypes.TOP else [0, 0, -configuration.bond_length]
+        )
+        passivant_bond_vector_crystal = material.basis.cell.convert_point_to_crystal(bond_vector)
+        return (np.array(surface_atoms_coordinates) + np.array(passivant_bond_vector_crystal)).tolist()
+
+
+class UndercoordinationPassivationBuilderParameters(BaseModel):
+    """
+    Parameters for the  UndercoordinationPassivationBuilder.
+    Args:
+        coordination_threshold (int): The coordination threshold for undercoordination.
+    """
+
+    cutoff: float = 3.0
+    coordination_threshold: int = 3
+
+
+class UndercoordinationPassivationBuilder(PassivationBuilder):
```

Comment:
The name is hard to understand - needs a better name

Reply:
CoordinationBasedPassivationBuilder